frontend/lib/resolution/scope-queries.cpp

```diff
@@ -2674,131 +2681,137 @@ std::vector<ID> findUsedImportedIds(Context* context,
+      if (p.second) {
+        // Insertion occured, so this is the first time visiting this scope.
+        // Gather the IDs of the used/imported modules
+        auto vec = findUsedImportedIds(context, scope);
+        for (const auto& id: vec) {
+          // filter out enums since they are not interesting here
+          if (parsing::idIsModule(context, id)) {
+            // save the module used/imported to ids
+            auto p2 = idSet.insert(id);
+            if (p2.second) {
+              // insertion occured, so add it also to the vector
+              idVec.push_back(id);
+            }
+          }
+        }
       }
     }
   }
 
-  // Recurse, but skip modules, since we do not consider their use/imports.
-  for (auto child : ast->children()) {
-    if (child->isModule()) continue;
-    findAllModulesUsedImportedInTreeImpl(context, ret, child->id());
+  // gather information about what use and import statements refer to
+  bool enter(const Use* d) {
+    processUseImport(d);
+    return false;
+  }
+  void exit(const Use* d) { }
+  bool enter(const Import* d) {
+    processUseImport(d);
+    return false;
+  }
+  void exit(const Import* d) { }
+
+  // do not delve into submodules
+  bool enter(const Module* m) {
+    return (m == inMod); // visit the requested module only
+  }
+  void exit(const Module* m) { }
+
+  // traverse through anything else
+  bool enter(const AstNode* ast) {
+    return true;
+  }
+  void exit(const AstNode* ast) { }
+};
+
+static const std::vector<ID>&
+findAllModulesUsedImportedInTreeQuery(Context* context, ID modId) {
+  QUERY_BEGIN(findAllModulesUsedImportedInTreeQuery, context, modId);
+  std::vector<ID> ret;
+
+  auto ast = parsing::idToAst(context, modId);
+  if (ast) {
+    if (auto mod = ast->toModule()) {
+      GatherUsedImportedIds gatherModules(context, mod);
+      ast->traverse(gatherModules);
+      // swap the gathered vector in to place
+      ret.swap(gatherModules.idVec);
+    }
   }
-}
 
-static const std::map<ID, ID>&
-findAllModulesUsedImportedInTreeQuery(Context* context, ID id) {
-  QUERY_BEGIN(findAllModulesUsedImportedInTreeQuery, context, id);
-  std::map<ID, ID> ret;
-  findAllModulesUsedImportedInTreeImpl(context, ret, id);
   return QUERY_END(ret);
 }
 
-static void moduleInitVisitModules(Context* context, ID idMod,
+static void moduleInitVisitModules(Context* context, ID modId,
                                    std::set<ID>& seen,
-                                   const ID* idTrigger,
-                                   std::vector<std::pair<ID, ID>>& out) {
-  CHPL_ASSERT(!idMod.isEmpty());
+                                   std::vector<ID>& out) {
+  CHPL_ASSERT(!modId.isEmpty() && parsing::idIsModule(context, modId));
 
   // Avoid a cycle walking modules we've seen before.
-  if (seen.find(idMod) != seen.end()) return;
+  if (seen.find(modId) != seen.end()) return;
 
   // We are at least visiting ourselves.
-  seen.insert(idMod);
+  seen.insert(modId);
```

Comment:
Isn't the 'insert-and-check' idiom applicable here?

Reply:
Good point, I will change it to do that.